packages/dds/tree/src/simple-tree/treeApi.ts

```diff
@@ -105,12 +111,41 @@ export const treeNodeApi: TreeNodeApi = {
 		// The parent of `node` is an object, a map, or undefined (and therefore `node` is a root/detached node).
 		return parentField.parent.key;
 	},
-	on: <K extends keyof FlexTreeNodeEvents>(
+	on: <K extends keyof TreeChangeEvents>(
 		node: TreeNode,
 		eventName: K,
-		listener: FlexTreeNodeEvents[K],
+		listener: TreeChangeEvents[K],
 	) => {
-		return getFlexNode(node).on(eventName, listener);
+		const flex = getFlexNode(node);
+		const anchor = flex.anchorNode;
+
+		switch (eventName) {
+			case "afterShallowChange": {
+				// The funky pattern subscribing to two events from the anchors is so we can fire afterShallowChange once,
+				// batching changes to several fields of the node. 'childrenChanged' on the anchor fires on every change to
+				// a field so doesn't allow us to batch.
```

Comment:
What's the motivation for batching? Depending on the answer, the doc on `afterShallowChange` (which doesn't say anything about any batching) may need to be updated.

Reply:
Just to keep the event less spammy. Since they're subscribing to events on the node, I think it feels more natural from the consumer's perspective if we can tell them "hey, one or more things changed on this node" just once during a delta visit, instead of a separate firing for each changed property of the node.

And I think this matches the current docs for `afterShallowChange` (emphasis on "one or more"):

> Emitted by a node when:
> - The content of **one or more** of the node's properties changes (i.e., the property now contains a new node, or
nothing if it previously contained a node).
> - [other cases]

Reply:
A bit similar to how `afterDeepChange` feels more natural firing once after all changes to the subtree have happened (with due consideration for the trickiness from double-pass), instead of once per subtree change.

Reply:
You could also implement this "funky pattern" as:

```ts
return anchor.on("childrenChanged", () => {
  const offSubtreeChanged = anchor.on("subtreeChanged", () => {
    listener();
    offSubtreeChanged();
  })
});
```

In addition to being more concise, it makes it more immediately clear that:
* "subtreeChanged" is dependent on "childrenChanged" firing first
* "subtreeChanged" is only expected to fire once per "childrenChanged"

Reply:
My original concern has been resolved, so as far as that is concerned, this comment thread can be resolved.
Get you own comment threads @noencke 😛

Reply:
I'm saving our team money. GH charges by the thread, but not by the comment.

Reply:
I can't tell if that's a joke or not. Is that how GH Enterprise charges? 😮

Reply:
@noencke had to go a bit more complicated than your suggestion for refactoring the funky pattern, but still ended up with something simpler than what I had :).

All issues in the thread resolved now!

Reply:
Nice!! I think if you wanted to make it one step simpler, you could have the `offSubtreeChanged` be the `let` binding in the upper scope, replacing `alreadySubscribedToSubtreeChanged`. If `offSubtreeChanged` is undefined, then you know the event hasn't been registered yet, and if it is defined, then you know it has (and you can also use it to deregister). Then you don't need the boolean at all.

Totally up to you if you like that. I am probably being too detail-oriented here, so don't feel like I am strongly opinionated; I just think it is a fun puzzle to reduce the code down :)